pkg/loop/internal/relayer/pluginprovider/contractreader/contract_reader.go

```diff
@@ -495,6 +545,54 @@ func (c *Server) QueryKey(ctx context.Context, request *pb.QueryKeyRequest) (*pb
 	return &pb.QueryKeyReply{Sequences: pbSequences}, nil
 }
 
+func (c *Server) QueryKeys(ctx context.Context, request *pb.QueryKeysRequest) (*pb.QueryKeysReply, error) {
+
+	var filters []types.ContractKeyFilter
+	for _, keyQuery := range request.Filters {
+
+		contract := convertBoundContractFromProto(keyQuery.Contract)
+
+		queryFilter, err := convertQueryFiltersFromProto(keyQuery.Filter, contract, c.impl)
+		if err != nil {
+			return nil, err
+		}
+
+		sequenceDataType, err := getContractEncodedType(contract.ReadIdentifier(queryFilter.Key), c.impl, false)
+		if err != nil {
+			return nil, err
+		}
+
+		filters = append(filters, types.ContractKeyFilter{
+			Contract:         contract,
+			KeyFilter:        queryFilter,
+			SequenceDataType: sequenceDataType,
+		})
+
+	}
+
+	limitAndSort, err := convertLimitAndSortFromProto(request.GetLimitAndSort())
+	if err != nil {
+		return nil, err
+	}
+
+	sequencesIter, err := c.impl.QueryKeys(ctx, filters, limitAndSort)
+	if err != nil {
+		return nil, err
+	}
+
+	sequences := make([]types.SequenceWithKey, 0)
+	for k, s := range sequencesIter {
+		sequences = append(sequences, types.SequenceWithKey{Key: k, Sequence: s})
+	}
+
+	pbSequences, err := convertSequencesWithKeyToVersionedBytesProto(sequences, request.Filters, c.encodeWith)
```

Comment:
Let's skip this collection step by passing the `iter.Seq` along:
```suggestion
	sequences, err := c.impl.QueryKeys(ctx, filters, limitAndSort)
	if err != nil {
		return nil, err
	}

	pbSequences, err := convertSequencesWithKeyToVersionedBytesProto(sequences, request.Filters, c.encodeWith)
```